witnet/witnet-rust
# feat(wallet): add runRadRequest /createMnemonics method handlers

# Implement `runRadRequest` and `createMnemonics` handlers

This PR implements the a JSON-RPC handler named `runRadRequest` and the other named `createMnemonics`, the former can be used to execute a RAD request while the other to generate a [Mnemonic Phrase](https://en.bitcoin.it/wiki/Seed_phrase) that will be used to derive the wallet addresses. 

For more information about what a RAD request is see the [Witnet docs](docs.witnet.io).

Related PRs: #584 

## How to test

1. Run the wallet with: `cargo run -p witnet_wallet -- run`
2. Open a browser web console and write:
```
var sock= (() => { let s = new WebSocket('ws://localhost:11212');s.addEventListener('message', (e) => {  console.log('Rcv =>', e.data) });return s; })();
sock.send('{"params":{"deliver":[{"url":"https://hooks.zapier.com/hooks/catch/3860543/l2awcd/","kind":"HTTP-GET"},{"url":"https://hooks.zapier.com/hooks/catch/3860543/l1awcw/","kind":"HTTP-GET"}],"consensus":{"script":[145,146,102,32]},"aggregate":{"script":[145,146,102,32]},"retrieve":[{"script":[152,83,204,132,146,1,163,98,112,105,204,132,146,1,163,85,83,68,204,132,146,1,170,114,97,116,101,95,102,108,111,97,116,204,130],"url":"https://api.coindesk.com/v1/bpi/currentprice.json","kind":"HTTP-GET"}],"not_before":0},"id":1,"method":"runDataRequest","jsonrpc":"2.0"}');
// result => {"jsonrpc":"2.0","result":{"Float":{"value":7918.545}},"id":1}

sock.send('{"jsonrpc":"2.0","method":"createMnemonics", "params": { "length": "Words12" },"id":"1"}');
// result => {"jsonrpc":"2.0","result":"pole live more ride great above junk goat scrub vacant outside receive","id":"1"}
```
You should see the following output in the server logs:

```
% cargo run -p witnet_wallet -- run
   Compiling witnet_wallet v0.2.0 (/home/anler/src/public/witnet-rust/wallet)
    Finished dev [unoptimized + debuginfo] target(s) in 10.76s
     Running `target/debug/witnet_wallet run`
[DEBUG] Loading config from `/home/anler/src/public/witnet-rust/witnet.toml`
[DEBUG] RAD-Executor started
[DEBUG] RAD-Executor started
[DEBUG] RAD-Executor started
[DEBUG] RAD-Executor started
[DEBUG] RAD-Executor started
[DEBUG] RAD-Executor started
[DEBUG] RAD-Executor started
[DEBUG] RAD-Executor started
[INFO ] Starting 1 workers
[INFO ] Starting server on 127.0.0.1:11212
[INFO ] Received connection from 127.0.0.1:4704
[TRACE] Received request: {"params":{"deliver":[{"url":"https://hooks.zapier.com/hooks/catch/3860543/l2awcd/","kind":"HTTP-GET"},{"url":"https://hooks.zapier.com/hooks/catch/3860543/l1awcw/","kind":"HTTP-GET"}],"consensus":{"script":[145,146,102,32]},"aggregate":{"script":[145,146,102,32]},"retrieve":[{"script":[152,83,204,132,146,1,163,98,112,105,204,132,146,1,163,85,83,68,204,132,146,1,170,114,97,116,101,95,102,108,111,97,116,204,130],"url":"https://api.coindesk.com/v1/bpi/currentprice.json","kind":"HTTP-GET"}],"not_before":0},"id":1,"method":"runRadRequest","jsonrpc":"2.0"}
[DEBUG] [ThreadId(2)] Executing RAD request...
[DEBUG] [ThreadId(2)] Finished RAD request.
[DEBUG] Sending response Object({"Float": Object({"value": Number(7932.64)})})
[TRACE] Sending response: {"jsonrpc":"2.0","result":{"Float":{"value":7932.64}},"id":1}
[TRACE] Received request: {"jsonrpc":"2.0","method":"createMnemonics", "params": { "length": "Words12" },"id":"1"}
[DEBUG] Handling request for method: createMnemonics
[DEBUG] Generating mnemonics with length: Words12
[TRACE] Sending response: {"jsonrpc":"2.0","result":"pole live more ride great above junk goat scrub vacant outside receive","id":"1"}
```

## Implementation details

```
     +-----------------------+      +-------------------------+
     |       (App Actor)     |      |  (RadExecutor actor)    |
     |                       |--+--->                         |
     | RunRadRequest handler |  |   | - executes the rad req  |
     +-----------------------+  |   |   in a sync context     |
                                |   | - retrievals are run    |
                                |   |   in parallel (rayon)   |
                                |   +-------------------------+
                                |
                                |
                                |   +-------------------------+
                                `--->  (RadExecutor actor)    |
                                    |        ...              |
                                    +-------------------------+
```

For each rad request, the request handler delegates the execution to a `RadExecutor` actor in a separate thread that will resolve the RAD request and return the appropriate response. The `RadExecutor` will also try to run the _retrievals_ of the RAD request in parallel.
